### ISSUE TITLE
Deprecate the async API and replace it with CLR abstractions for threading

### ISSUE THREAD

**Issue by RobertBouillon** [dotnet/runtime#45159]
# Overview
The current [asynchronous programming model](https://docs.microsoft.com/en-us/dotnet/csharp/async) is unnecessary because __all__ code written in .NET is asynchronous by nature of running on a multi-threaded operating system. If we don't require [explicit language support](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/async/) for executing on a system-managed thread, we shouldn't require [explicit language support](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/async/) for executing in a CLR-managed context. It should be the job of the CLR, not the developer, to determine the context in which a method is executing and which context control method is required (e.g. `Thread.Sleep` vs `Task.Delay`) _when_ it is required.

```csharp
public static int Program()
{
  IFoo foo = Foo.Create();
  return foo.Bar();
}

public class Sync : IFoo
{
  int IFoo.Bar() => 100;
}

public class Async : IFoo
{
  int IFoo.Bar()
  {
    wait 1000;  //Thread.Sleep(1000) if the context is a thread, Task.Delay(1000) if the context is a Task.
    return 50;
  }
}
```

## Problem Statement
I understand the implications of breaking changes, and I understand the initial resistance to deprecating major functionality. In this case, I believe there is more harm in retaining support for the async model than in removing it. The async model is a minefield of 'gotchas' that make the language less stable and less accessible.

1. The [current model](https://docs.microsoft.com/en-us/dotnet/csharp/async) violates fundamental OO design principles.
  The [TAP pattern](https://docs.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/task-based-asynchronous-pattern-tap) requires that the interface describe the implementation. The async model requires that interfaces not only describe what members an object has, but also how those members behave (whether or not they block / yield). The result is that a truly agnostic interface must have both a sync and and async method variant for __every method__ (See example at bottom). This can further complicate interface design because properties, despite being methods "under-the-hood", [cannot be async](https://github.com/dotnet/roslyn/issues/3015). Thus, properties cannot be used in truly-agnostic interfaces, and must be replaced with Get/Set methods which each have sync and async variants. An interface, [by definition](https://www.cs.utah.edu/~germain/PPS/Topics/interfaces.html#:~:text=In%20Object%20Oriented%20Programming%2C%20an,and%20a%20turn_off()%20method.), should only need to describe what an object does, not how. Whether or not the member blocks / yields is a matter for the documentation, not the method signature. That's the way it's always been for multi-threading, and I see no benefit to the new model.

2. The async model makes C# less accessible to less-experienced programmers
The intuitive way to retrieve the result of an async operation from a synchronous context is `Task.Result`. (But, wait, do you use `Task.Result` or `Task.GetAwaiter().Result`?) This can [often lead to deadlocks](https://stackoverflow.com/questions/17248680/await-works-but-calling-task-result-hangs-deadlocks), and internet searches are either [fantastically technical](https://blog.stephencleary.com/2012/07/dont-block-on-async-code.html) or [ambiguously simple](https://stackoverflow.com/questions/27265818/does-the-use-of-async-await-create-a-new-thread), because with async there appears to be no in-between. The fact is that the current model is simply not easy to use, and is easy to do 'wrong'.

3. The async model makes C# makes threading more difficult for experienced programmers.
In complex systems that utilize the async model, thread synchronization can become an issue when accessing critical sections. C# provides a plethora of low-level [thread synchronization primitives](https://docs.microsoft.com/en-us/dotnet/standard/threading/overview-of-synchronization-primitives) that have evolved since the early days of programming. Many effective, well-established patterns exist around these primitives. Using these primitives in an async context can result in undefined behavior as existing primitives can block the thread instead of yielding the task.

# Proposed Solution
The [design objectives for the async model](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/async/) are already possible with operating system's thread scheduler (traditional threading). Context switches are expensive, however, and presumably this gave rise to the [TaskScheduler](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskscheduler?view=net-5.0). The concept of a `TaskScheduler` in lieu of a managed `ThreadScheduler` makes sense, because a task may not run on a thread (It could be [Overlapped I/O](https://docs.microsoft.com/en-us/windows/win32/fileio/synchronous-and-asynchronous-i-o?redirectedfrom=MSDN), a [message pump](https://devblogs.microsoft.com/pfxteam/await-and-ui-and-deadlocks-oh-my/), etc.).

So if the objective is to allow code to run in an asynchronous context, why not use the API that has successfully enabled multi-threading out of the OS? Why create a new one? If .NET seeks to create a higher-level abstraction of the execution context, that should be a function of the CLR, __not__ the application logic. We have a virtual machine - we should use it, and this is a great use-case.

## Proposal # 1: Refactor the existing Threading API
The existing thread API (`Thread.Sleep`, `Thread.Yield`, etc) maps directly to operating-system threads. Instead, these can be made context-aware. If the current context is a `Task`, calling `Thread.Sleep` would call `Task.Delay`. If the current context is an OS Thread, `sleep` would be called on the OS thread. `Thread.Yield` would call `Task.Yield` or pass `sleep(0)` to the OS, so-on and so-forth. The same refactoring would need to take place for all threading "primitives," including mutexes, semaphores, wait handles, etc.

Benefits:
- Fewer changes to the C# API and language spec
- Familiar API for existing
- Backwards compatibility with complex threading in pre-async applications.

Drawbacks:
- Lose direct access to OS threads

## Proposal # 2: Create a new API that abstracts context control and synchronization
Create a new object model and / or C# language features that provide context control to the application, regardless of what the context may be (A thread, Task, etc). Thread / Task control would all be performed through the single API.

Benefits:
- Purpose-Specific API would allow C# to define optimal behaviors for context control and synchronization, setting a new standard for languages to abstract context control on a VM

Drawbacks:
- Significant design effort

### C# language support 
```csharp
public void Synchronous()  //Current 
{
  Thread.Yield();
}

public async Task Asycnchronous()  //Current 
{
  await Task.Yield();
}

public void Agnostic()  //Proposed
{
  yield;
}
```

### Example
This is my program. This is what I _need_ to do. The logic doesn't need to (and shouldn't) change depending on whether the file is on a local drive (and the calls synchronous) or the file is on Azure (and the calls are asynchronous). The interface defines what my program cares about. A file needs a name, and I need to open it. This is the case with traditional threading. The implementation should not change depending on whether or not this is running on a Thread or in a Task (async context).

```csharp
public void Process(IEnumerable<IFile> files)
{
  foreach (var textFile in files.Where(x => x.Name.EndsWith(".txt")))
    using (var stream = textFile.Open())
      ReadTextFile(stream);
}
```
This is what my interface would like, normally.

```csharp
public interface IFile
{
  string Name { get; set; }
  Stream Open();
}

public class AzureBlob : IFile
{
  public string Name
  {
    get => GetFileName();  //Async call
    set => SetFileName(value);  //Async call
  }

  public Stream Open() => OpenFile();  //Async call
}

public class OsFile : IFile
{
  FileInfo _file;
  public string Name
  {
    get => _file.Name;
    set => File.Copy(_file.Name, value);
  }

  public Stream Open() => _file.OpenRead();
}
```

This is the abomination my code becomes when the [TAP pattern](https://docs.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/task-based-asynchronous-pattern-tap) is applied to the interface. Each method in each concrete class implements either the async or sync method, and simply creates a redirection from the alternate method. This creates a lot of duplicate code and is certainly not the most efficient design paradigm, both from the perspective of performance _and_ maintainability.

```csharp
  public interface IFile
  {
    string GetName();
    Task<string> GetNameAsync();

    void SetName(string name);
    Task SetNameAsync(string name);

    Stream Open();
    Task<Stream> OpenAsync();
  }

  public class AzureBlob : IFile
  {
    public void SetName(string name) => Task.Run(() => SetNameAsync(name));
    public async Task SetNameAsync(string name) => await SetFileName(name);

    public string GetName() => Task.Run(GetNameAsync).Result;
    public async Task<string> GetNameAsync() => await GetFileName();

    public Stream Open() => Task.Run(OpenFile).Result;
    public async Task<Stream> OpenAsync() => await OpenFile();

  }
  public class OsFile : IFile
  {
    FileInfo _file;
    public string GetName() => _file.Name;
    public Task<string> GetNameAsync() => Task.FromResult(_file.Name);

    public Stream Open() => _file.OpenRead();
    public Task<Stream> OpenAsync() => Task.FromResult((Stream)_file.OpenRead());

    public void SetName(string name) => _file.CopyTo(name);
    public Task SetNameAsync(string name) => await SetName(name);
  }
```

**Comment by Clockwork-Muse** [comment]
> I guess the first question I'd like help answering is this - can the CLR do it? Can the CLR abstract the thread context? Is there something I'm missing about async/await, or perhaps the CLR, that makes what I'm suggesting impossible?

If we ignored the existence of `async`/`await` (and probably a few other things in native interop code), you're looking at doing something like Java's [Project Loom](https://cr.openjdk.java.net/~rpressler/loom/loom/sol1_part1.html).  So the basic answer is "the engineering is possible in the general sense".

> This is my program. This is what I need to do. The logic doesn't need to (and shouldn't) change depending on whether the file is on a local drive (and the calls synchronous) or the file is on Azure (and the calls are asynchronous).

Generally _all_ IO, including reading from local drives, should be asynchronous, just because of how slow touching a drive is (this used to matter far more before the advent of modern SSDs).  This is before you consider things like network drives, where you might **think** the file is local, but the OS is just helpfully presenting a file somewhere out on the internet.  Or silly situations where you have a really slow local drive and a really fast internet connection, and it's faster to read from the internet than the local drive.

> Why must we force developers to write the same procedure in two different ways depending on the context? 

... generally you don't.  Outside of a few specific things that are best done asynchronously (like IO), you write a synchronous implementation.  Which you can wrap into a `Task` if the situation warrants it.  There are very few things that should be explicitly using `Thread` in modern C#.

**Comment by RobertBouillon** [comment]
> Generally all IO, including reading from local drives, should be asynchronous

One implementation of an [abstract File API](https://github.com/RobertBouillon/Pillars/blob/main/Spin.Pillars/FileSystem/File.cs) I have represents a [virtual file system](https://gist.github.com/RobertBouillon/6b14e61306e3d3c4d0e4f83d890f08ec) built on the browser's [local storage](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage) and accessed [via Blazor](https://gist.github.com/RobertBouillon/8bf41f0def62d8f035a40d36d0d3a3dc). It is synchronous because it is marshalled in-memory. Unit tests running on .NET 5 provide the file as a [local system file](https://github.com/RobertBouillon/Pillars/blob/main/Spin.Pillars/FileSystem/OS/OsFile.cs), which is asynchronous, while the runtime application running on MONO (Blazor) in the browser provides a file from the browser, which is synchronous. Since a File is really just a named blob, I would not want to impose a synchronicity on the API, which is abstract. Instead, I'm forced to require that concrete classes of my abstract class implement both synchronous and asynchronous variants.

This highlights my first objection to async/await - we require that the method signature describe the implementation. This is a constraint imposed by the implementation, not the requirements. This is backwards and is a reason to deprecate the async/await API. If we were to look at the OO design of C# in a vacuum, and were to ask the question, "What should be defined in a method signature?", would we really say that we should specify the synchronicity? I would posit "no" as all methods should be presumed asynchronous; any code could be suspended at any time. The code should be indifferent to whether it yields execution to the OS or the CLR. Yet async methods have distinct signatures from their synchronous counterparts, which has a profound impact in API design.

**Comment by dylech30th** [comment]
I think you are confusing the "asynchronous" with "multithreading", in C#, `Task` is not a sort of "replicate" of `Future`, and "asynchronous" is also working in a single thread environment, every `async`-marked-method which contains `await` modifier will be transformed into a state machine(you can check it through [Sharplab.io](https://sharplab.io/#v2:EYLgHgbALAPgsAKAAICYAMikGYAEq8CMA7DgN6I6V65IER4AchArAHRL0CyAhgJYB2ACgCUZClQlIAnDgCavAKYAbACYiA3OMoBfLTj3ZC9JE1psOcxapFiEEyTLPsIrACLLuAT0FY0f4Zp29pTSLM5uHt6+/oHBIY4E5i7uSl4+fmgBeroI2kA=)), where the original method is split into several "stages". If you remove `async`-`await`, there's no way a compiler knows if any method should be transformed into a state machine or not, the only way is to transform every method which...definitely would come with a high cost. So if you want to use `async/await` and TAP, I suggest you to learn some coroutine concepts instead of keep using the multithreading way, the `SynchronizationContext` is actually a `Dispatcher` which dispatch some task to run on a specified thread, and the so-called "context switching" here, might not be the thread context switching.
I've also noticed those two links you've mentioned above, [Don't block on async code](https://blog.stephencleary.com/2012/07/dont-block-on-async-code.html) and [Awaits and UI and deadlocks oh my](https://devblogs.microsoft.com/pfxteam/await-and-ui-and-deadlocks-oh-my/), the scenarios mentioned in these two posts are also a matter of `SynchronizationContext`, because the default behaviors of `SynchronizationContext` in UI applications are dispatched its continuations to original thread, so you cannot block to wait for the result on UI thread while the `SynchronizationContext` trying to dispatch its result back to its captured context(you can consider as the thread by which invokes `await` operation, at here, it's UI thread), which apparently will causing deadlocks, and that's why `ConfigureAwait(false)` appears — it tells the awaiting `Task` to not use the default behavior(dispatch to original thread), but dispatch the continuations to another thread, and then deadlock will definitely disappear, and if you understand what does "Dispatch" mean here, those bugs would be clear for you
Also, I cannot find any requirement to write two methods in your examples, because one `SetNameAsync ` or `GetNameAsync` or `OpenAsync` would be enough because you can just left the `Task` there and do not apply the `await` modifier and it will not suspended, and if you really want internal logic to be pure synchronous, you can just use `Task.FromResult`, it's born for this—And first of all, like what @Clockwork-Muse said: All I/O operations should be asynchronous:
```cs
static async Task Main(string[] args)
{
    var azure = new AzureFile();
    var local = new LocalFile();
    var stopWatch = new Stopwatch();
    Console.WriteLine("Start");
    stopWatch.Start();
    azure.OpenFileAsync();
    Console.WriteLine("azure.OpenFileAsync(): This invocation neither suspended nor blocked, it will just run like a background task, because you are not using await, and its internal logic is asynchronous: " + stopWatch.Elapsed.Seconds);
    stopWatch.Restart();
    await azure.OpenFileAsync();
    Console.WriteLine("await azure.OpenFileAsync(): This invocation actually gets suspended, but not blocked, because it is modified by await operator: " + stopWatch.Elapsed.Seconds);
    stopWatch.Restart();
    _ = azure.OpenFileAsync().Result;
    Console.WriteLine("azure.OpenFileAsync().Result: This invocation actually gets blocked, because you are accessing Task.Result, warning, this operation may cause deadlock in some specified SynchronizationContext: " + stopWatch.Elapsed.Seconds);
    stopWatch.Restart();
    await local.OpenFileAsync();
    Console.WriteLine("await local.OpenFileAsync(), This invocation actually gets blocked despite of await operator, because it's internal logic is blocked and the Task is just a wrapper of the result: " + stopWatch.Elapsed.Seconds);
    stopWatch.Restart(); 
    local.OpenFileAsync();
    Console.WriteLine("local.OpenFileAsync(): This invocation actually gets blocked despite of await operator, because it's internal logic is blocked and the Task is just a wrapper of the result: " + stopWatch.Elapsed.Seconds);
    stopWatch.Restart();
    _ = local.OpenFileAsync().Result;
    Console.WriteLine("local.OpenFileAsync().Result: This invocation actually gets blocked, and it's safe because the Task is already completed when return: " + stopWatch.Elapsed.Seconds);
}

interface IFile
{
    Task<Stream> OpenFileAsync();
}

class AzureFile : IFile
{
    // You want it to run asynchronously? not problem
    public async Task<Stream> OpenFileAsync()
    {
        await using var memo = new MemoryStream();
        await new FileInfo("some file") /* we are simulating here so I just use FileInfo */ .OpenRead().CopyToAsync(memo);
        await Task.Delay(3000);
        return memo;
    }
}

class LocalFile : IFile
{
    // You want the internal logic to be completely synchronous? it's OK
    public Task<Stream> OpenFileAsync()
    { 
        using Stream memo = new MemoryStream();
        new FileInfo("some file") /* we are simulating here so I just use FileInfo */ .OpenRead().CopyTo(memo);
        Thread.Sleep(3000);
        return Task.FromResult(memo);
    }
}

// outputs:
// Start
// azure.OpenFileAsync(): This invocation neither suspended nor blocked, it will just run like a background task, because you are not using await, and its internal logic is asynchronous: 0
// await azure.OpenFileAsync(): This invocation actually gets suspended, but not blocked, because it is modified by await operator: 3
// azure.OpenFileAsync().Result: This invocation actually gets blocked, because you are accessing Task.Result, warning, this operation may cause deadlock in some specified SynchronizationContext: 3
// await local.OpenFileAsync(), This invocation actually gets blocked despite of await operator, because it's internal logic is blocked and the Task is just a wrapper of the result: 3
// local.OpenFileAsync(): This invocation actually gets blocked despite of await operator, because it's internal logic is blocked and the Task is just a wrapper of the result: 3
// local.OpenFileAsync().Result: This invocation actually gets blocked, and it's safe because the Task is already completed when return: 3
```

**Comment by En3Tho** [comment]
Well, I do not see comparisons with Go have much value because the lack of async-await model in Go. So you can't just make an async method mimic sync method in Go. But you can mimic Goroutines in .Net, you have Channels, RX, Data flow, MailboxAgent (F#). I think what you really propose is first class language support for something like Goroutines maybe, something to simplify creating reactive non direct-awaiting API's? RX comes to mind again but it has a learning curve and it's not out of the box.
 I think it's true that in C# we always have to operate with 'runtime types' like Task instead of some language-specific abstraction. Same with Actions and Funcs. So maybe this is really a language suggestion and not a runtime one?

**Comment by RobertBouillon** [comment]
> method which contains `await` modifier will be transformed into a state machine

Correct - what I'm suggesting is that this be a function of the CLR / VM, not the compiler / language. In the example you provide, async/await seems not only benign, but arguably superfluous. This is because you're async "all-the-way-down." That is, you're starting in an async context with `static async Task Main`. In my experience, most async code looks identical to sync code, except Task/Async methods are decorated with awaits. The problems arise when you mix sync / async code, and this can happen a lot for a lot of reasons, but especially with shared libraries, where the best approach is to provide "overloads" for both calling conventions.

So if async works best if it's _everywhere_ or _nowhere_, why do we need it at all? What's the difference between "legacy" code that blocks and awaited code, if they function the same (giving the impression of serial execution)? The only difference is how the program is yielding control while waiting for data: with sync code, yielding is a context switch, and with async code, it's a state machine an execution remains in the process. The only difference, and the reason for the existence of async/await, is performance - avoiding the context switch.

We need to pull this back to the root cause, the fundamental problem that async/await tries to solve. All code is a series of instructions executed sequentially. At some point, usually some form of I/O, program execution cannot continue because it does not have the data required. So what does the CPU _DO_ then? Traditionally, the OS suspends the thread by storing its context: registers, stack, etc, so the CPU can start processing work from another program. The thread will be resumed when the data is available. This context switch is expensive. The state machine attempts to do the exact same thing more efficiently without ceding the quantum and incurring a context switch. Ergo the only purpose async/await serves is to avoid a context switch: like killing a fly with a bazooka.

The default mode-of-operation should be sequential / synchronous. This is, after all, what `await` tries to emulate. Any parallelization should be explicit. To my third point in the OP, the async/await model falls apart with complex code that includes critical sections. It's fine if all you're doing is `await` with a couple `Task.WaitAny` and `Task.WaitAll` for parallelization. If you want something like a [ReaderWriterLock](https://docs.microsoft.com/en-us/dotnet/api/system.threading.readerwriterlock?view=net-5.0), you're asking for trouble because this will block the execution, at worst deadlocking, and at best you're back to where you started with context switches.

All methods should be blocking, by default. Again, this is what `await` emulates. What I'm suggesting is that when a method "blocks", the CLR can perform an optimized switch to another Task or cede the quantum to the OS if it has no work to do. The solution we need is more efficient context switching, and that should be a function of the CLR.

**Comment by dylech30th** [comment]
It is true that if CLR can automatically switch the execution context when the CPU is idle, then `async/await` will become meaningless,  but as we all know, `async/await`' always know where to return when `await` ends thanks to the state machine, in your proposal, CLR should be able to switch to another context automatically, then consider this scenario: CLR post an I/O task to another thread, and that I/O job also post another I/O task to the third thread, now we have three threads running three tasks concurrently, and the problem is: how does CLR cooperate these related tasks? Since there are no longer any state machines (obviously you cannot generate them at runtime or it will become a profoundly complicated task because you need to modify not only the callee but also every callers' code and turn them into state machines, this will bring a huge time complexity let alone the difficulty of its implementation, I don't think JIT can afford this), you might need an extra call stack that goes across all the threads to store the context on each time it switches which seems still cannot achieve what you said "more efficient context switching" because pop and push operations might be very fast but an extra stack also brings extra space costs

**Comment by Clockwork-Muse** [comment]
> All methods should be blocking, by default. Again, this is what `await` emulates. What I'm suggesting is that when a method "blocks", the CLR can perform an optimized switch to another Task or cede the quantum to the OS if it has no work to do. The solution we need is more efficient context switching, and that should be a function of the CLR.

Again, what you seem to be asking for is Java's Project Loom.

> The threading model is the concern of the application, not the library.

For ease-of-use reasons, most applications (and their programmers) would **gladly** cede concerns about threading to the library for things that actually require concurrent execution, not just asynchronous execution.

<hr />

Speaking specifically about your file api example, if it was rooted in "traditional" IO and just happened to have a Blazor frontend I would just do the implementation async anyways.  `Task`s are **allowed** to complete synchronously, and immediately, and a number of libraries perform just this optimization (and the runtime can perform a bunch of its own optimizations for some other cases).  If you wanted to really push the envelope, return `ValueTask` instead of `Task`.
If it's primarily intended as some sort of api just over the browser storage, using a real file backing store doesn't make much sense for unit tests; you should be using some sort of in-memory store even there, especially since hitting against a real filesystem has concurrency issues for tests run in parallel.

**Comment by RobertBouillon** [comment]
> Again, what you seem to be asking for is Java's Project Loom.

I appreciate the reference. I wasn't aware of this project before and I agree that it looks very similar to what I'm proposing. Do you have any thoughts on the applicability of Project Loom to .NET? Are you aware of any previous discussions on the topic? Do you have any insight you can provide on that project?

> most [programmers] would gladly cede concerns about threading to the library

I liken this to how few libraries are "thread-safe". It's generally accepted that threading and context management is handled by the application (with a few exceptions, of course).

> Speaking specifically about your file api

The solution you propose follows the exact train of thought that spawned this post: it seems the best way to avoid problems with async is to make **everything** async, even stuff that *might* complete synchronously, in case it *might* execute asynchronously. That being the case, why mark any interface as synchronous? As a thought experiment, assume we don't. Okay now if everything's async, why do we need any special language features? Why are we passing `Task` instead of `void` everywhere? Well we don't need the language features: the compiler/CLR can handle this. But wait, we're right back where we started! C# without async/await! So what does the language feature do if it's not needed? It spawns the generation of a state machine. Well, if the OS can switch without needing a generated state machine, why should we? Why not just have the VM do what the OS is doing? We already have "managed" threads and Tasks. And here we are. 

>  consider this scenario [...] how does CLR cooperate these related tasks?

That's getting too detailed for where we're at right now, I think. I'm more concerned with approval of the concept than the "how." I believe it is possible, but I'm hoping to leverage the collective experience of the community and SME's here to vet whether or not we "should" at this point. If you believe what I'm asking is technically impossible or inefficient in-practice, could you please elaborate?

**Comment by Clockwork-Muse** [comment]
> I appreciate the reference. I wasn't aware of this project before and I agree that it looks very similar to what I'm proposing. Do you have any thoughts on the applicability of Project Loom to .NET? Are you aware of any previous discussions on the topic? Do you have any insight you can provide on that project?

It was decided (At the time `async`/`await` was first designed, and referenced in some more recent discussions) that the explicit use of a syntax element to call attention to the continuation/coroutine/etc was preferable.  That is, it was explicitly decided to **not** go the route that Loom is currently exploring/attempting (although Loom itself is more recent, the basic syntax/runtime ideas are not).  It's extremely unlikely that this decision will be revisited or changed.

**Comment by RobertBouillon** [comment]
I appreciate the feedback - it's very helpful.

I think the difference between when the decision was made and now is that we have the benefit of experience - we now have a full view of the benefits and drawbacks of async/await. The deeper I dig into async/await, the more it occurs to me that the only tangible benefit to async/await is performance - specifically avoiding an OS context switch. Everything that can be done *with* async/await can be done *without* it, assuming we permit liberal use of context switching. There is a significant amount of change in the API and patterns for the sake of avoiding context switching. IMO, there are far more drawbacks to async/await then benefits. When a solution creates more problems than it solves (the cure is worse than the disease), it should be revisited.

So I feel compelled to succinctly restate the top 3 drawbacks to async/await:

1. The execution context (sync/async) is now part of the method signature, effectively dividing all code into two categories.
2. High learning curve for novice engineers (easy to deadlock, ambiguous API members, lots of patterns to learn)
3. Limitations to advanced threading - unable to use threading primitives in async contexts with no alternatives.

This seems like a lot to take on for a bump in performance by avoiding context switches. Is there not a better way to avoid context switching? Am I missing some other tangible benefits? Is my assessment of the limitations incorrect?

**Comment by Clockwork-Muse** [comment]
> 1. The execution context (sync/async) is now part of the method signature, effectively dividing all code into two categories.

AKA [what color is your function](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/).  However most code _does_ generally only fall into one bucket or the other.  Like I said before, though, there was an explicit decision to make this bucketing clear.

> 2. High learning curve for novice engineers (easy to deadlock, ambiguous API members, lots of patterns to learn)

While it's true that it might be easy to _misuse_ the API and deadlock, correctly using the API is simple, and for basic use there's only one pattern to learn.

> 3. Limitations to advanced threading - unable to use threading primitives in async contexts with no alternatives.

I'm not sure what you mean by "advanced threading".  It's true that you can't use `lock`, `Monitor`, and a few other things in an async context, but there are some alternatives depending on what effect you were actually trying to do

**Comment by RobertBouillon** [comment]
I believe we see the severity of the drawbacks differently. Can we agree then that the aforementioned caveats, at best, introduce complexity, and at worst impact usability and stability? That is, if we could accomplish the same benefit that async/await provides (fast context switching) without any of the caveats, wouldn't that be superior?

**Comment by John0King** [comment]
I'm wondering it that possible to `sync-wait` a async method without  first "block" the current thread.

first 
```C#
async Task FooAsync()
{
    await File.ReadAllBytesAsync();
    await File.ReadAllBytesAsync();
    await File.ReadAllBytesAsync();
}

```

```C#
Task Main()
{
    //thread 1
    await FooAsync();  // thread   on os or somewhere
}
// there are  only 1 thread  running if there are all async await.

void Main(){
   // thread 1
   // block
    FooAsync().GetAwaiter().GetReault(); thread 2, the first `await` in this method cause it use a new threadPool thread and current thread is not been released
}
// there are 2+ thread
``` 

can we just say we want to execute(or wait ) all code on current thread (eg.  transform  `async` to `sync` )?

```C#
void Main()
{
    sync_wait FooAsync();
}
```

some ideas 

```
foreach( var action in state-machine)
{
     fat-call action(); // if it contains state-machine code (eg. async)  fat the state-machine like this method.
    // if it need, we create just one thread (event loop) for one application to handle some notify-complete call
}
```

**Comment by HaloFour** [comment]
@RobertBouillon 

You aren't eliminating the caveats, you're replacing one set for a different set.  What Project Loom and Goroutines does is hide that asynchrony behind imperative code and cooperative multitasking of green threads, and it is very slick, but it's also bound by management of OS threads and the preemptive multitasking of the OS.  Those ecosystems need to be _very_ carefully groomed so that _every_ possible case that might block on I/O can recognize that it is running on such a context and execute a different path to yield the green thread, establish a polling/notification mechanism to know when that operation is complete and then allow the OS thread to be scheduled elsewhere.  None of this is automatic and the OS is totally unaware that this is happening at all.  Because the green threads are backed by a finite number of worker OS threads _any_ blocking code can cause a severe impact to performance or even deadlocks.

I'm honestly surprised about Java Loom.  Java is making a gamble here that the number of places within the runtime is limited enough that they can make the required changes so that any blocking code will instead yield.  When Project Loom is released every bit of native code that blocks will have to be similarly updated.  This is really only feasible because the amount of native code is relatively small and limited to very specific frameworks managed by relatively big players.  It's my opinion that the amount of native code in the .NET world is orders of magnitude greater due to the ease in which .NET can interop with other languages and frameworks, something actively discouraged in the Java world.

I've also spoken about Loom to members of the C# team and their opinion is that asynchrony is something that the developer should be very explicit about and that the behavior of automatically yielding or blocking depending on your context would be undesirable.

**Comment by Clockwork-Muse** [comment]
> I'm wondering it that possible to `sync-wait` a async method without first "block" the current thread.
> 
> first
> 
> ```cs
> async Task FooAsync()
> {
>     await File.ReadAllBytesAsync();
>     await File.ReadAllBytesAsync();
>     await File.ReadAllBytesAsync();
> }
> ```
> 
> ```cs
> Task Main()
> {
>     //thread 1
>     await FooAsync();  // thread   on os or somewhere
> }
> // there are  only 1 thread  running if there are all async await.
> 
> void Main(){
>    // thread 1
>    // block
>     FooAsync().GetAwaiter().GetReault(); thread 2, the first `await` in this method cause it use a new threadPool thread and current thread is not been released
> }
> // there are 2+ thread
> ```

Note that a `Task Main()` actually performs the `task.GetAwaiter.GetResult()` call, so your two programs are equivalent.  That aside, the number of threads is somewhat non-deterministic - if the file is small enough (and the drive is fast enough), you might only ever have one thread, and the entire program run synchronously.  `await` is not `Task.Run()` - using it doesn't start a new thread, at all; what it does do is say "if for some reason this isn't complete yet, remember where I was so somebody can deal with it later".  On a modern system that's _usually_ a separate thread, but that's not a requirement.

> 
> can we just say we want to execute(or wait ) all code on current thread (eg. transform `async` to `sync` )?
> 
> ```cs
> void Main()
> {
>     sync_wait FooAsync();
> }
> ```

The catch is that `async`/`await` code shows its best advantage when there's nothing for the current thread to execute - things like web requests, etc, where a relatively large chunk of time is better spent doing _anything_.

**Comment by John0King** [comment]
@Clockwork-Muse  I know `async/await` is great, but  the "async all the way down" is not.
for example, that async alreay make `event` not useful  (eg. `event Func<Task> OnDoSomting`);
and it also trouble for api design, for example  like the mvc filter, we know that the  action "may" be async
so 
```
intercept before method()
await action();
intercept after method();
```

but  can we know that for intercepter before and after is sync ?  no, it may also have async call,  so

```
await intercepter before methodAsync()
await action()
await intercepter after methodAsync()
```

and this is just 1 example that we facing it (I know that for mvc there is a async filter solve this problem perfectly ) ,  
and this make our api everything async,  and when we design a class, we normally facing:    1) does the method need return `Task`,  2) does the callback need to be a `Func<Task>` .

I think that why many people ask for things like goLang. 

I'm hoping that we can have way to maybe use only extra 1 thread  to solve the problem that   `sync -> async -> sync -> async` that use more than 3 threadPool thread with each call and cause threadpool starvation when height concurrent  and  eventually cause really really bad performance on our app.

**Comment by RobertBouillon** [comment]
@HaloFour I'm not clear on what, specifically, you object to. You present a very critical case against Project loom. To be clear, I am not proposing we emulate Project Loom. @Clockwork-Muse had simply pointed out some similarities in the objectives.

Async/await exists primarily to avoid context switching, and does so with generated state machines. It accomplishes this with a large number of serious caveats which damage the language. I believe we can avoid context switching while rebuffing the caveats introduced with async/await.

I am suggesting that the framework (CLR/Compiler) manage the context of the thread, not the developer. If the current context is a state machine, yield to the state machine. If the current context is an OS thread, yield to the OS thread. In either case, the developer need only state the intention: `yield`. The developer is still in complete control of the thread's behavior.

**Comment by HaloFour** [comment]
@RobertBouillon 

> Async/await exists primarily to avoid context switching

No, async/await exists to avoid blocking OS threads on I/O bound tasks.  You can't avoid context switching.

What you are describing _is_ an implementation of Project Loom on the .NET runtime.  There's no other way for this "yield" that you would describe to work.  You either block an OS thread, or you don't.  Not blocking an OS thread requires an implementation of green threads, which are _still_ backed by OS threads.  There's no such thing as "yielding to a state machine".  The closest you get is parking a green thread.

**Comment by RobertBouillon** [comment]
> No, async/await exists to avoid blocking OS threads on I/O bound tasks. 

A context switch is *how* the OS thread "blocks." So we're saying the same thing here. And it doesn't *need* to be I/O bound - that's just the typical use-case. For example, one thread might block while waiting for another thread to finish.

> You can't avoid context switching.

Yes you can, and that's what async/await is designed to do. That's what green threads are designed to do. I don't know enough about Loom to know if green threads are truly as difficult or pointless as you claim, but I'm certain they're not the "only" way to avoid context switching, as is demonstrated by the existence of async/await (and [Interlocked](https://docs.microsoft.com/en-us/dotnet/api/system.threading.interlocked.increment?view=net-5.0), for that matter).

I've worked on high-performance systems that are very sensitive to context-switching. We did a lot to avoid them. The design goal is to keep the CPU loaded with as much "real work" as possible. It sounds like green threads are something worth looking into, however I'd be happy if we could just abstract the async context so I don't have to write the same method two different ways (one for sync, and one for async). Too often I have to write a Method and AsyncMethod, which have largely the same body, with `await` scattered across the latter. And I'd love to use threading primitives again.

**Comment by John0King** [comment]
just to point out that   typescript/javascript is not these kind annoying  is because
1)  it only have one reference type `Promise`
2)  dynamic type

```ts
OnBeforeMethodExecute(): boolean | Promise<boolean>;
OnAfterMethodExecute(): boolean | Promise<boolean>;

// use  
async callMethodAsync() : void
{
    let b = OnBeforeMethodExecute();
    let br = true;
    if(b instanceof Promise)
    {
        br = await b;
    }
    else
    {
        br = b;
     }
     console.log(b) // do something
    await method();


   // ...  same for after
}

```

**Comment by HaloFour** [comment]
@RobertBouillon 

> however I'd be happy if we could just abstract the async context so I don't have to write the same method two different ways

That design in intentional, the teams don't want this "implicitness" in asynchrony as offered by Go or Java Loom.  They want it to be obvious that you are crossing an asynchronous boundary and are explicitly opting into the costs of doing so.  

I'd love to see coroutine primitives offered by the runtime so that I could write my own, but that wouldn't at all change how `async`/`await` works in the language.  You'd never get to the point where a method can be asynchronous and return a non-task-like type.  The only way to enable that is green threads.

I'll let @CyrusNajmabadi weigh in on this as he and I have discussed it several times.

**Comment by Clockwork-Muse** [comment]
@John0King 
 
```javascript
OnBeforeMethodExecute(): boolean | Promise<boolean>;
OnAfterMethodExecute(): boolean | Promise<boolean>;

// use  
async callMethodAsync() : void
{
    let b = OnBeforeMethodExecute();
    let br = true;
    if(b instanceof Promise)
    {
        br = await b;
    }
    else
    {
        br = b;
     }
     console.log(b) // do something
    await method();


   // ...  same for after
}
```
C# equivalent:

```csharp
async Task callMethodAsync()
{
    // Assuming OnBeforeMethodExecute is Func<object>
    var b = OnBeforeMethodExecute();
    var br = true;
    if (b is Task<bool> c)
    {
        br = await c;
    }
    else if (b is bool d)
    {
        br = d;
    }

    Console.WriteLine(br);
    await method();
}
```

... note that in many cases I'd just make the before/after methods _always_ `async`, and just use `Task.FromResult(<whatever>)` if I didn't do any asynchronous work - there would be no state machine transitions, and the entire thing would have run synchronously.

> I'm hoping that we can have way to maybe use only extra 1 thread to solve the problem that sync -> async -> sync -> async that use more than 3 threadPool thread with each call and cause threadpool starvation when height concurrent and eventually cause really really bad performance on our app.

.... I'm not sure why you believe this uses _more_ than 3 threadpool threads; if you're dealing with IO, it's probably only blocking two.  Note that a `sync -> async` call means that you're doing `task.Result` or `task.Wait()` (or similar) which should almost always be avoided.

**Comment by John0King** [comment]
@Clockwork-Muse  the problem is that in C#  any type can be await, including `string`  `bool` and so on.   (stand type now have 4 type `Task`  `Task<T>`  `ValueTask` `ValueTask<T>` , and for ValueTask you can not using `is ValueTask`  because `ValueTask<T>` is not `ValueTask`)
2nd ,  the object will cause value type (`struct`) to box into heap,  and that why people want to avoid.

I hope `Shap` and `Union` type come in , and support things like typescript.

**Comment by RobertBouillon** [comment]
Having spent significant time researching this issue (thanks to everyone who commented), I've taken the long-way around to the same conclusion as @HaloFour - the only alternative to async/await is green threads. All of the caveats I've mentioned can be traced back to some dependency on the state machine, meaning that to solve the caveats, we have to go back to stack manipulation for context changes.

So then my question goes back out to the group - why not green threads? I've spent some time looking at Java, Rust, and Golang literature on green threads, and it looks like the big hang-up seems to be a question of optimization. I don't see anything that really says green threads are "bad," mostly that they're "hard." But it seems like it would move the burden of complexity from the software to the framework, which seems like the right move. 

Async/await seems to serve two main functions: avoid context switching and simulating synchronous code for coroutines.

I'm skeptical of the former, as single memcpy to/from the L1 cache (i.e. context switch) seems more efficient than random access to the heap (state-machine fields). This could lead to better controlled benchmarks, but is this really better in an environment with resource-contention (e.g. a VM or 80%-utilization server)? And the coroutine candy seems to be a net loss, where not having to maintain a string of callbacks requires significant compromises to software and API design. Personally, I'd rather manage the callbacks explicitly than deal with all the caveats of await/async.

So I'd like to close this issue - I think we can dress up async/await all we want, but you can only put so much lipstick on a pig. Given the benefit of hindsight, I really think the right way moving forward is green threads. Before I close this, though, I'm just curious about what other people think, and if anyone knows the current temperature on the .NET team regarding green threads?

**Comment by HaloFour** [comment]
I've spoken to some members of the C# language design team on this issue, particularly pertaining to my experiences with the OpenJDK Loom early access builds, and the impression that I got was that while green threads are interesting and that there are proponents to that strategy that ultimately the design of being very intentional about async boundaries won over.  In a green thread world, as in a normal blocking world, it can be very hard to tell when you're crossing that boundary.  It also necessitates a different set of APIs for fork/join and other patterns, which kind of come for free if you have to `await` the awaitable.

There's also the issue that green threads still need to be scheduled and run by proper OS threads.  You need total cooperation from within the ecosystem to "park" the green thread, establish some kind of notification and then allow that OS thread to move on to something else.  Without that, a blocking call will block both the green thread and the OS thread on which it has been scheduled, which brings you right back to square one.  Java is making the gamble that they can control enough of the locations in the JVM where blocking can occur that this won't be an issue, but they can't handle it everywhere (DNS lookup on Windows comes to mind) and all bets are off with native code.  In .NET I think that's a substantially more difficult problem as mixing .NET with native is so much easier and much more common, and the runtime delegates to the native OS libraries in more circumstances.  It would probably be way too easy to accidentally block all of the OS worker threads and cause all of the green threads to come screeching to a halt.

**Comment by RobertBouillon** [comment]
This issue has been superseded by #50796
